### PR TITLE
Refactor old

### PR DESCRIPTION
**Pull request by r0mainK**
Hey !

So first off thanks for the repo, it really helped having a tested CRF in pytorch. I spent the past couple cycles improving your code, and thought I would PR. Here is the list of changes I PR ed here (high level, we can talk of each commit during the review):

- remove useless methods: you had a lot of methods needlessely encapsulating code, I removed them as it made it harder to read. Specifically, I removed `reset_parameters` because it is like creating a new layer, I removed `_compute_score` and `_compute_normalize` to have only one loop, and because separating thr code made it harder to understand, and I removed `_viterbi_decode` because it was useless
- refactor validation: `ValueError` and `assert` were competing for validating inputs, but there was a ton of redundance and ValueError were more informative, so I removed all the asserts. I also added a check on the mask, and moved some code around for efficiency
- optimize decoding: I made the second part of the viterbi decoding in batch, and now return a tensor. this allows for an important speedup (even for small batch sizes). It's better to never convert back to list in code, it should be upt to the user to do this conversion

There is also some minor changes, like adding pytorch to the setup.py. I think the commits are well split so it should be no problem to review. I tested on my computer with pytorch 1.0, 1.1 and 1.4, all work. 

PS: I tried to keep your coding style as much as possible, but I would recommend removing the line comments. Your code is good enough so that is impedes comprehension in my opinion.

EDIT: Added optimization of numerator computation

### CODE REVIEW

**Comment by kmkurn** [comment]
Hi, thank you for the PR! Really appreciate it. However my hands are a bit full at the moment, and this is not a small PR, so it might take a while before I can review. But from skimming the code:
- I prefer that the version is not bumped (yet) because I'd rather do that when I'm about to release the next version
- Please bring back `reset_parameters` to keep it aligned with PyTorch's convention where each layer has that method (which is still the case last time I checked)

Thanks again for the PR!

**Comment by r0mainK** [comment]
No problem, I'm not in a rush :)

I'll remove both commits, indeed I hadn't seen that the `reset_parameters` was idiomatic, I had checked only the superclass `nn.Module`  but not the actual instances !

File: torchcrf/__init__.py [hunk 1]
```diff
@@ -204,20 +208,22 @@ def decode(self, emissions: torch.Tensor,
             mask = mask.transpose(0, 1)
         self._validate(emissions, mask)
 
-        # emissions: (seq_length, batch_size, num_tags)
-        # mask: (seq_length, batch_size)
```
Comment (kmkurn) [inline torchcrf/__init__.py]: Please keep these comments.

File: torchcrf/__init__.py [hunk 2]
```diff
@@ -204,20 +208,22 @@ def decode(self, emissions: torch.Tensor,
...
         seq_length, batch_size = mask.shape
 
+        # score is a tensor where for every batch, value at column j stores the score of the
+        # best tag sequence so far that ends # with tag j
```
Comment (kmkurn) [inline torchcrf/__init__.py]: Remove the extra `#`

File: torchcrf/__init__.py [hunk 3]
```diff
@@ -226,50 +232,41 @@ def decode(self, emissions: torch.Tensor,
...
 
-            # Reverse the order because we start from the last timestep
-            best_tags.reverse()
-            best_tags_list.append(best_tags)
+        # Trace back where the best last tag comes from, and add to our sequences
+        for i in range(seq_length - 1):
+            best_prev_tag = history[i, torch.arange(batch_size), best_prev_tag]
+            best_tags[:, seq_length - 2 - i] = best_prev_tag
+        # Add the tag for the last timestep
+        best_tags[torch.arange(batch_size), mask.long().sum(dim=0) - 1] = best_last_tag
```
Comment (kmkurn) [inline torchcrf/__init__.py]: Shouldn't both `torch.arange` be called with `device=best_tags.device`?

File: torchcrf/__init__.py [hunk 4]
```diff
@@ -226,50 +232,41 @@ def decode(self, emissions: torch.Tensor,
...
-        best_tags_list = []
-
-        for idx in range(batch_size):
-            # Find the tag which maximizes the score at the last timestep; this is our best tag
-            # for the last timestep
-            _, best_last_tag = score[idx].max(dim=0)
-            best_tags = [best_last_tag.item()]
+        # Now, compute the best tag sequences for each sample
+        # shape: (batch_size, seq_length)
+        best_tags = emissions.new_empty(batch_size, seq_length, dtype=torch.long)
```
Comment (kmkurn) [inline torchcrf/__init__.py]: Could you please use `new_zeros` here? The tag IDs must be between 0 and `num_tags` to avoid error if they're to be used by the caller.

File: torchcrf/__init__.py [hunk 5]
```diff
@@ -102,23 +102,30 @@ def forward(
...
         # shape: (batch_size,)
         numerator = self.start_transitions[tags[0]]
         numerator += emissions[0, torch.arange(batch_size), tags[0]]
-        denominator = self.start_transitions + emissions[0]
+        # Transition score and emission to next tag, only added if next timestep is valid
+        # shape: (batch_size,)
+        numerator += ((
+            emissions[torch.arange(1, seq_length).view(-1, 1).repeat(1, batch_size),
+                      torch.arange(batch_size).repeat(seq_length - 1, 1), tags[1:]] +
+            self.transitions[tags[:-1], tags[1:]]) * mask[1:].float()).sum(dim=0)
```
Comment (kmkurn) [inline torchcrf/__init__.py]: This looks rather complicated. I prefer the old one for readability, could you please revert it back?
